### PR TITLE
Port `StringToArray` to `function-arrays` subcrate

### LINKED ISSUE #9497: Port `StringToArray` to `function-arrays`
### Is your feature request related to a problem or challenge?

It is part of https://github.com/apache/arrow-datafusion/issues/9285. 
It is a good first issue. ArrayToString is already done in https://github.com/apache/arrow-datafusion/pull/9113, so you can easily find out which files are needed to modify.

#9352 #9425 are also nice for reference

### Describe the solution you'd like

Port StringToArray.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by erenavsarogullari**
## Which issue does this PR close?
Closes #9497.

## What changes are included in this PR?
This PR aims to do following changes  in terms of Epic #9285:
1- Port `StringToArray` to `function-arrays` subcrate,
2- Argument length check is added to `string_to_array` function: [ref](https://github.com/apache/arrow-datafusion/pull/9543/files#diff-ee636e21e786e49965f9c86552d090386b3174d558528e715dc98df2a93d3796R262)
3- Case: delimiter is `NULL` was breaking test case under `array.slt` such as `SELECT string_to_array('abc', NULL)`. This case is also handled: [ref](https://github.com/apache/arrow-datafusion/pull/9543/files#diff-5c50694a837ad80e51ea9a289bd43ac59eee372404dd0caefe714ecae3a103a9R145)
4- Function Documentation is added.

## Are these changes tested?
Yes, all `array.slt` based `string_to_array` tests are passed.

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/core/tests/dataframe/dataframe_functions.rs [hunk 1]
```diff
@@ -870,3 +870,22 @@ async fn test_fn_array_to_string() -> Result<()> {
 
     Ok(())
 }
+
+#[tokio::test]
+async fn test_fn_string_to_array() -> Result<()> {
+    let expr = string_to_array(lit("abc##def##ghi"), lit("##"), lit("!"));
```
Comment (jayzhan211) [inline datafusion/core/tests/dataframe/dataframe_functions.rs:876]: Some test cases are already covered in `array.slt`

File: docs/source/user-guide/sql/scalar_functions.md [hunk 2]
```diff
@@ -3113,6 +3113,7 @@ _Alias of [make_array](#make_array)._
 ### `string_to_array`
 
 Splits a string in to an array of substrings based on a delimiter. Any substrings matching the optional `null_str` argument are replaced with NULL.
+`SELECT string_to_array('abc##def', '##')` or `SELECT string_to_array('abc def', ' ', 'def')`
```
Comment (jayzhan211) [inline docs/source/user-guide/sql/scalar_functions.md:3116]: 👍

**Comment by jayzhan211** [comment]
You can follow the `How to format .md document` section in https://arrow.apache.org/datafusion/contributor-guide/index.html#developer-s-guide to solve the formating issue

File: datafusion/core/tests/dataframe/dataframe_functions.rs (hunk 1 above)
Comment (erenavsarogullari) [reply datafusion/core/tests/dataframe/dataframe_functions.rs]: Firstly, thanks for review. Addressed.

**Review by jayzhan211 — APPROVED** [review]
Thanks @erenavsarogullari

**Pull request merged**